File: javascript/leetCodeChallenges/Easy/valid_parentheses.js
var isValid = function(s) {
    let stack = [];
    const pairs = {
        ')': '(',
        '}': '{',
        ']': '['
    };

    for (let char of s) {
        if (char === '(' || char === '[' || char === '{') {
            stack.push(char)
        } else {
            if (stack.length === 0 || stack.pop() !== pairs[char]) {
                return false
            }
        }
    }

    return stack.length === 0

};

File: python/LeetCodeChallenges/Easy/valid_parentheses.py

class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        pairs = {
            '(': ')',
            '{': '}',
            '[': ']'
        }

        for char in s:
            if char in pairs:
                stack.append(char)
            elif char in pairs.values():
                if not stack or pairs[stack.pop()] != char:
                    return False
            else:
                return False

        return not stack

# TIme complexity is O(n) and Space is O(n) in worst scenario